### DISCUSSION TITLE
[Bug]: Error translating local repository: js_md5__WEBPACK_IMPORTED_MODULE_2__ is not a function

### DISCUSSION THREAD

**Discussion by xybei** [babel/babel#14734] [Q&A] [answered]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

package.json
```json
{
  "name": "lib1.js",
  "version": "1.0.0",
  "description": "test project1",
  "main": "lib1.js",
  "dependencies": {
    "@babel/runtime-corejs3": "^7.18.3",
    "js-md5": "file:../js-md5"
  },
  "devDependencies": {
    "@babel/core": "^7.18.2",
    "@babel/plugin-transform-runtime": "^7.18.2",
    "@babel/preset-env": "^7.18.2",
    "babel-loader": "^8.2.5",
    "webpack": "^4.39.3",
    "webpack-cli": "^3.3.8"
  },
  "scripts": {
    "build": "webpack --mode development",
  }
}
```
I have made a sample project to reproduce this problem：[https://github.com/xybei/test_babel_loader_error.git](https://github.com/xybei/test_babel_loader_error.git)
```lang-js
// var md5 = require('js-md5');
// import md5 from 'js-md5';
// import { md5 } from 'js-md5';
import * as md5 from 'js-md5';

class Lib1 {
    static test() {
        const hash = md5("Message to hash");
        console.log(hash);

        const hash2 = md5.create();
        hash2.update('Message to hash');
        console.log(hash2.hex());
    }
}

export default Lib1;
```

### Configuration file name

babel.config.js

### Configuration

babel.config.js
```js
module.exports = {
    presets: [
        ["@babel/preset-env",
            {
                targets: {
                    ie: 10
                },
                // useBuiltIns: "usage",
                // corejs: 3,
                debug: true
            }
        ]
    ],
    plugins: [
        [
            "@babel/plugin-transform-runtime", {
                corejs: 3
            }
        ]
    ]
};
```

### Current and expected behavior

I want to use the js-md5 library as a local repository, but I get the following error：
> When I put the 'js-md5' repository as a local dependency:
```json
  "dependencies": {
    "@babel/runtime-corejs3": "^7.18.3",
    "js-md5": "file:../js-md5"
  },
```
- babel compiles with a warning:
**WARNING in ./src/lib1.js 18:18-28
"export 'create' (imported as 'md5') was not found in 'js-md5'**
- Error when running in browser::
**Uncaught TypeError: js_md5__WEBPACK_IMPORTED_MODULE_2__ is not a function**

### Environment

  System:
    OS: Windows 10 10.0.19044
  Binaries:
    Node: 14.15.4 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.10 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    @babel/core: ^7.18.2 => 7.18.2
    @babel/plugin-transform-runtime: ^7.18.2 => 7.18.2
    @babel/preset-env: ^7.18.2 => 7.18.2
    @babel/runtime-corejs3: ^7.18.3 => 7.18.3
    babel-loader: ^8.2.5 => 8.2.5
    webpack: ^4.39.3 => 4.46.0

### Additional context

> When I install the 'js-md5' repository normally from npmjs.com everything works fine
```json
  "dependencies": {
    "@babel/runtime-corejs3": "^7.18.3",
    "js-md5": "^0.7.3",
  },
```
> When I put the 'js-md5' repository as a local dependency, and remove ‘babel-loader’ config from webpack.config.js

- There are no warnings, the packaged file runs fine
```js
const path = require('path');
module.exports = {
    devtool: 'source-map',
    entry: './src/lib1.js',
    output: {
        path: __dirname,
        filename: `lib1.out.js`,
        library: 'Lib1',
        libraryTarget: 'umd',
        libraryExport: 'default',
        globalObject: 'this',
    },
    // module: {
    //     rules: [
    //         {
    //             test: /\.js$/,
    //             exclude: /node_modules/,
    //             use: {
    //                 loader: "babel-loader"
    //             }
    //         }
    //     ]
    // },
    resolve: {
        alias: {
            '@babel/runtime-corejs3': path.resolve(__dirname, './node_modules/@babel/runtime-corejs3')
        }
    }
};
```

[a comment by an automated account is not shown]

**Comment by nicolo-ribaudo** [comment, marked as answer]
By default Babel assumes that your files are ESM modules, but js-md5 is a CommonJS module and thus Babel "breaks" it (it adds some ESM syntax to that file, so then webpack reads it as ESM and cannot see the CommonJS exports). You can fix this in a few ways:
- Don't compile `md5-js` with Babel. You can do it by using `excludes: [/node_modules/, /js-md5/]` in the `babel-loader` config in `webpack.config.js`
- Add `sourceType: "unambiguous"` do your Babel config. This means "if a file contains `import`/`export` it's ESM, otherwise it's CommonJS": Babel would thus correctly recognize js-md5 as a CommonJS file, and it would transform it correctly.
- Add this to your Babel config:
  ```js
  overrides: [{
    test: /js-md5/,
    sourceType: "script"
  }]
  ```
  which explicitly tells Babel that `js-md5` is a CommonJS file and not an ESM module.

**Reply by xybei** [reply]
Thank you very much, all three methods work, I use ```exclude: [/node_modules/, /js-md5/]```, the packaged file is much smaller.
